src/platform/amd/acp_7_0/lib/clk.c

```diff
@@ -0,0 +1,492 @@
+	float fract_part    = 0.0f;
+	float final_refclk  = 0.0f;
+
+	boot_ref_clk_startup.u32all =
+		acp_reg_read_via_smn(CLK7_CLK_PLL_REFCLK_RATE_STARTUP, sizeof(int));
+
+	clk7_spll_field_9_t  clk_spll_field_9;
+
+	clk_spll_field_9.u32all = 0x0;
+	clk_spll_field_9.u32all = acp_reg_read_via_smn(CLK7_SPLL_FIELD_9, sizeof(int));
+
+	if (clk_spll_field_9.bitfields.spll_dpll_cfg_3 == 0x2)
+		final_refclk = ((32768.0f * 128.0f) / 1000000.0f);
+	else
+		final_refclk = (float)boot_ref_clk_startup.bitfields.main_pll_ref_clk_rate_startup;
+
+	clk7_clk_pll_req.u32all = acp_reg_read_via_smn(CLK7_CLK_PLL_REQ, sizeof(int));
+	clk7_spll_field.u32all = acp_reg_read_via_smn(CLK7_SPLL_FIELD_2, sizeof(int));
+
+	spinediv = (1 << clk7_clk_pll_req.bitfields.pllspinediv);
+
+	if (clk7_spll_field.bitfields.spll_fracn_en == 1)
+		fract_part = (float)(clk7_clk_pll_req.bitfields.fbmult_frac / (float)65536.0f);
+
+	*boot_ref_clk = (float)(((final_refclk) * (clk7_clk_pll_req.bitfields.fbmult_int +
+					fract_part)) / (float)spinediv);
+}
+
+void acp_change_clock_notify(uint32_t clock_freq)
+{
+	volatile clk7_clk1_dfs_cntl_u_t dfs_cntl;
+	volatile clk7_clk1_bypass_cntl_u_t bypass_cntl;
+	volatile clk7_clk1_dfs_status_u_t dfs_status;
+	volatile uint32_t updated_clk = 0;
+	float did = 0.0f;
+	float fraction_val = 0.0f;
+	uint32_t int_did_val = 0;
+	float boot_ref_clk;
+	acp_clock_type_t clock_type = acp_aclk_clock;
+
+	acp_7_0_get_boot_ref_clock(&boot_ref_clk);
+
+	tr_info(&acp_clk_tr, "acp_change_clock_notify clock_freq : %d clock_type : %d",
+		clock_freq, clock_type);
+
+	fraction_val = (float)(clock_freq / (float)1000000.0f);
+	clock_freq   = (clock_freq / 1000000);
+	if (acp_aclk_clock == clock_type) {
+		bypass_cntl.u32all = acp_reg_read_via_smn(CLK7_CLK1_BYPASS_CNTL, sizeof(int));
+		dfs_cntl.u32all = acp_reg_read_via_smn(CLK7_CLK1_DFS_CNTL, sizeof(int));
+	} else if (acp_sclk_clock == clock_type) {
+		bypass_cntl.u32all = acp_reg_read_via_smn(CLK7_CLK0_BYPASS_CNTL, sizeof(int));
+		dfs_cntl.u32all = acp_reg_read_via_smn(CLK7_CLK0_DFS_CNTL, sizeof(int));
+	}
+
+	bypass_cntl.bitfields.CLK1_BYPASS_DIV = 0;
+
+	if (clock_freq == 6 || clock_freq == 0) {
+		did = 128;
+		dfs_cntl.bitfields.CLK1_DIVIDER = 0x7F;
+		bypass_cntl.bitfields.CLK1_BYPASS_DIV = 0xF;
+	} else {
+		did  = (float)(boot_ref_clk / (float)fraction_val);
+		tr_info(&acp_clk_tr, "acp_change_clock_notify CLK Divider : %d boot_ref_clk : %d\n",
+			(uint32_t)(did * 100), (uint32_t)boot_ref_clk);
+
+		if (did > 62.0f) {
+			dfs_cntl.bitfields.CLK1_DIVIDER = 0x7F;
+		} else {
+			fraction_val = did - (uint8_t)(did);
+			did = did - fraction_val;
+			if (did <= 16.00f)
+				did = (did * 4.0f);
+			else if ((did > 16.0f) && (did <= 32.0f))
+				did = ((did - 16.0f) * 2.0f + 64.0f);
+			else if ((did > 32.0f) && (did <= 62.0f))
+				did = ((did - 32.0f) + 96.0f);
+
+			int_did_val = (uint32_t)(fraction_val * 100.0f);
+			fraction_val = (float)(int_did_val / 100.0f);
+
+			if (fraction_val == 0.0f)
```

Comment:
I thought this wouldn't work, but after a brief test it does appear to work. Still, I'm not sure this would work always, maybe better do `-1e-10 < x < 1e-10` or similar instead of `== 0`

Reply:
Even if some rounding error "breaks" it for some values, similar rounding errors can affect `x < 1e-10` too.

0.01 does not exist in `float`, it's more like: 0.00999999977648258209228515625
https://numeral-systems.com/ieee-754-converter/

In any case lines 227 and 228 are confusing and seem slower and unnecessary.